ENH: Add built-in function for Styler to format the text displayed for missing values

immaxchen opened the pull request:
As described in the issues, user who wants to control how NA values are printed
while applying styles to the output will have to implement their own formatter.
(so that the underlying data will not change and can be used for styling)

Since the behavior is common in styling (for reports etc.), suggest to add this
shortcut function to enable users format their NA values as something like '--'
or 'Not Available' easily.

EDIT:

- Change implementation to integrate with the original `.format()` using `na_rep` argument
- Add a new table-wise default `na_rep` setting, which can be set through `.set_na_rep()`
- Minor fix: formatter should be wrapped outside of locs-loop 
- Added a few user guide examples and test cases

example usage: `df.style.highlight_max().format(None, na_rep="-")`

- [x] closes #21527 and #28358
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

immaxchen committed:
Add built-in funcion for Styler to format the text displayed for missing values

As described in GH #28358, user who wants to control how NA values are printed
while applying styles to the output will have to implement their own formatter.
(so that the underlying data will not change and can be used for styling)

Since the behavior is common in styling (for reports etc.), suggest to add this
shortcut function to enable users format their NA values as something like '--'
or 'Not Available' easily.

example usage: `df.style.highlight_max().format_null('--')`

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -110,6 +110,7 @@ Other enhancements
 - :meth:`DataFrame.to_json` now accepts an ``indent`` integer argument to enable pretty printing of JSON output (:issue:`12004`)
 - :meth:`read_stata` can read Stata 119 dta files. (:issue:`28250`)
 - Added ``encoding`` argument to :func:`DataFrame.to_html` for non-ascii text (:issue:`28663`)
+- :meth:`Styler.format_null` is now added into the built-in functions to help formatting missing values (:issue:`28358`)
```
can u add this into the user guide as well

topper-123 replied:
I'd like the name ``format_nans`` better, to be similar to ``fillna``, ``hasnans`` etc.

@jreback?

immaxchen replied:
Sure, I'll be glad to!

jbrockmendel commented:
@topper-123 My understanding is that the "Style" tag is refers to code style, linting stuff, not the Styler class.  Do I have this backwards?

immaxchen committed:
Add Styler.format_null into user_guide and reference Doc

immaxchen commented:
The doc has been added, please have a look! :smile:
How about the method name? do we have a conclusion?

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -930,6 +930,25 @@ def hide_columns(self, subset):
+        -------
+        self : Styler
+        """
+        self.format(
```
This looks like it will overwrite the formatting of a previously applied formatter for non-NA values. Something like

```python
df.style.format("hi-{}".format).format_null()
```

is that the case?

immaxchen replied:
Thanks @TomAugspurger, I like the name `format_na`! and yes, I was intended to make an overwriting implementation. Actually, I have considered the `set_*` approach as you, but it seems confusing for the case:
```
df.style.format('{:.2%}').set_na_format('-') # got 'nan%' instead of '-'
```
I've got a new idea, how about interface like this?
```
.format_na('-', subset=['col1','col2'])
.format('{:.2%}', na_rep='-', subset=['col3','col4'])
```
And the docstring for `format_na` rephrase to:
```
Format the text display value using default formatter but represent nan as `na_rep`.
For more advanced formatting, use Styler.format() with your custom formatter.
```

TomAugspurger replied:
I may not have been clear about my concern. It's fine that `na_format` overwrites the formatting for NA values. I'm concerned tht it overwrites the formatting for *non*-NA values. In my `.format("hi-{}".format).format_na('NA')` example, the NA values should be formatted as 'NA' and the non-NA values should be formatted as `hi-<value>`. But I suspect that right now the non-NA formatting is lost (though perhaps it's not).

TomAugspurger replied:
Not sure about adding an `na_rep` to the `.format` function... That's probably fine. I think it'd still be useful for users to have a way to control the default NA formatting at the table level.

But if we add an `na_rep` to `format`, then we wouldn't need a new `format_na` method, right?

immaxchen replied:
Sounds good, so a setting at the table level: `Styler.na_rep`
and the `def format(self, formatter, subset=None):` becomes 
`def format(self, formatter=None, na_rep=None, subset=None):`
drop `.format_na('-')`, use `.format(na_rep='-')` instead, right?

TomAugspurger replied:
I think that sounds correct. I'm not sure what the default should be, but probably just None (no special formatting for NA values).

TomAugspurger reviewed (commented):
We've been moving away from null and to NA. So I think this should be called `format_na` if anything.

I'm a bit concerned about the implementation. I suspect this would be better handled as an argument to `Styler` and a method like `set_na_format` (see how uuid is handled)`. Then the `default_display_func` would check if the value is NA and use the na format if it's NA.

immaxchen committed:
Revised to change implementation to integrate with the original Styler.format method

Change the implementation to integrate with the original `.format()` method by `na_rep` parameter
Add a new table-wise default `na_rep` setting, which can be set through the new `.set_na_rep()` method
Also enhanced the `.highlight_null()` method to be able to use `subset` parameter
Add a few user guide examples and test cases

immaxchen committed:
resolve merge conflicts

immaxchen committed:
add more tests for styling with NA values

immaxchen commented:
Revision done, please help to review, thanks!  :D

immaxchen commented:
I think this PR can also resolve #21527 (ENH: Styler display of NaN / null Values)

TomAugspurger commented:
Most likely not for a couple weeks.

On Tue, Oct 29, 2019 at 9:31 AM Max Chen <notifications@github.com> wrote:

> Hi @TomAugspurger <https://github.com/TomAugspurger>, would you be
> available to review this PR? thanks! 😉
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/29118?email_source=notifications&email_token=AAKAOIR4YVQVJA2WYY3VZPDQRBCNZA5CNFSM4JCVRYEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECQWUEY#issuecomment-547449363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAOISPN5GHSCPUW5BP4YLQRBCNZANCNFSM4JCVRYEA>
> .
>

immaxchen commented:
Ouch. :joy: 
@jreback can you help to review? :grin:

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -416,16 +423,20 @@ def format_attr(pair):
         )
 
-    def format(self, formatter, subset=None):
+    def format(self, formatter=None, subset=None, na_rep=None):
```
What was the point of changing `formatter` here?

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -891,6 +908,23 @@ def set_table_styles(self, table_styles):
         self.table_styles = table_styles
         return self
 
+    def set_na_rep(self, na_rep):
```
Can you annotate this function? I think should be

```suggestion
    def set_na_rep(self, na_rep: str) -> "Styler":
```

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -935,19 +969,21 @@ def _highlight_null(v, null_color):
         )
 
-    def highlight_null(self, null_color="red"):
+    def highlight_null(self, null_color="red", subset=None):
```
How is this related to `na_rep`?

immaxchen replied:
Actually not related. :joy: 
Just making it consistent with `highlight_min` and `highlight_max`, also for completeness of missing values styling and formatting.

WillAyd replied:
Nice idea but would prefer if you take this out then. We try to keep PRs to changing one thing at a time as it helps speed up review process

WillAyd replied:
Can always do as a follow up PR

immaxchen replied:
Thanks, that is a great advice!

immaxchen committed:
Merge remote-tracking branch 'upstream/master' into styler-builtin-shortcut-for-missing-values-formatting-gh28358

sync with the latest upstream/master

immaxchen committed:
revision based on the requested changes

1. keep formatter as mandatory in `.format` method
2. annotate the new method `.set_na_rep`
3. remove changes in `.highlight_null` to another PR
4. minor refinement to the whats new and user guide

immaxchen commented:
Revision done, please have a look. :grin:

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -71,6 +71,11 @@ class Styler:
         The ``id`` takes the form ``T_<uuid>_row<num_row>_col<num_col>``
         where ``<uuid>`` is the unique identifier, ``<num_row>`` is the row
         number and ``<num_col>`` is the column number.
+    na_rep : str or None, default None
```
I think this should be `str, optional`.

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -1480,3 +1518,13 @@ def _maybe_wrap_formatter(formatter):
         raise TypeError(msg)
+
+
+def _maybe_wrap_na_formatter(formatter, na_rep):
```
Can this be folded into `_maybe_wrap_formatter`? I'm finding the multiple layers of wrapping hard to follow.

immaxchen replied:
No problem, all done!

TomAugspurger reviewed (commented):
Thanks. Can you add

1. A test exercising passing na_rep to Styler, `Styler(df, na_rep="-")`
2. A test ensuring that non-numeric NA values are handled correctly? At least NaT for datetime data.

immaxchen committed:
add tests, enhance doc string and formatter wrapping

immaxchen committed:
Merge remote-tracking branch 'upstream/master' into styler-builtin-shortcut-for-missing-values-formatting-gh28358
sync with the latest upstream/master

immaxchen commented:
Hi Tom, William, please have a look at the revision when you are available, thanks~

immaxchen committed:
resolve merge conflict

WillAyd commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -114,6 +114,8 @@ Other enhancements
 - Added ``encoding`` argument to :func:`DataFrame.to_html` for non-ascii text (:issue:`28663`)
 - :meth:`Styler.background_gradient` now accepts ``vmin`` and ``vmax`` arguments (:issue:`12145`)
+- :class:`Styler` added :meth:`Styler.set_na_rep` method to set default missing values representation for the entire table.
+  :meth:`Styler.format` added the ``na_rep`` parameter to help format the missing values (:issue:`21527`, :issue:`28358`)
```
I think just this note is fine (can delete the line above)

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -126,6 +131,7 @@ def __init__(
         caption=None,
         table_attributes=None,
         cell_ids=True,
+        na_rep=None,
```
Can you annotate this? I think Optional[str]

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -416,16 +425,22 @@ def format_attr(pair):
         )
 
-    def format(self, formatter, subset=None):
+    def format(self, formatter, subset=None, na_rep=None):
```
annotation here as well

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -1487,14 +1523,22 @@ def _get_level_lengths(index, hidden_elements=None):
 
 
-def _maybe_wrap_formatter(formatter):
+def _maybe_wrap_formatter(formatter, na_rep):
```
and here

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -1487,14 +1523,22 @@ def _get_level_lengths(index, hidden_elements=None):
+    elif isinstance(na_rep, str):
+        return lambda x: na_rep if pd.isna(x) else formatter_func(x)
+    else:
+        msg = "Expected a string, got {na_rep} instead".format(na_rep=na_rep)
```
Can you add a test case that hits this?

immaxchen committed:
add type-hint and xfail test

immaxchen commented:
After adding the annotation for `na_rep`, mypy complained as below, so I have no choice but to add more annotations.
```
pandas/io/formats/style.py:136: error: Need type annotation for 'ctx'
pandas/io/formats/style.py:137: error: Need type annotation for '_todo' (hint: "_todo: List[<type>] = ...")
pandas/io/formats/style.py:158: error: Need type annotation for 'hidden_columns' (hint: "hidden_columns: List[<type>] = ...")
pandas/io/formats/style.py:173: error: Need type annotation for '_display_funcs'
pandas/io/formats/style.py:470: error: "None" not callable
```

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -126,9 +131,10 @@ def __init__(
-        self.ctx = defaultdict(list)
-        self._todo = []
+        self.ctx = defaultdict(list)  # type: DefaultDict[Tuple[int, int], List[str]]
+        self._todo = []  # type: List[Tuple[Callable, Tuple, Dict]]
```
Is it possible to add sub-types for Callable, Tuple, Dict here?

WillAyd commented on pandas/io/formats/style.py:
```diff
@@ -416,16 +427,22 @@ def format_attr(pair):
         )
 
-    def format(self, formatter, subset=None):
+    def format(self, formatter, subset=None, na_rep: Optional[str] = None):
```
Not required here but if you wanted to in a follow up annotate other arguments would be super helpful

WillAyd commented on pandas/tests/io/formats/test_style.py:
```diff
@@ -1009,6 +1009,75 @@ def test_bar_bad_align_raises(self):
+    def test_format_with_bad_na_rep(self):
+        # GH 21527 28358
+        df = pd.DataFrame([[None, None], [1.1, 1.2]], columns=["A", "B"])
+        df.style.format(None, na_rep=-1)
```
Instead of marking this as fail should use pytest.raises as a context manager (you'll find other examples throughout tests)

WillAyd reviewed (changes requested):
Looks good. Minor issue with the new test rest of the comments are optional and can be done as follow ups

immaxchen committed:
revise test using pytest.raises

immaxchen committed:
Merge remote-tracking branch 'upstream/master' into styler-builtin-shortcut-for-missing-values-formatting-gh28358

immaxchen commented:
Thanks @WillAyd, I have revised using `pytest.raises` and resolve merge conflict. Since this PR is getting quite lengthy, I will try to add annotations in the follow-ups.

WillAyd reviewed (changes requested):
OK last edit - I know you want to refine types further later but in the interim can you convert to the Py36 syntax? We have #29741 outstanding which is doing that for existing types, so want to use that in new PRs as well

immaxchen committed:
using py36 syntax for annotations

immaxchen commented:
Hi @WillAyd, revision done for py36 type syntax, please have a look :)

WillAyd reviewed (approved):
Nice PR this lgtm. @TomAugspurger care to look?

TomAugspurger commented:
Thanks @immaxchen!
